internal/controller/operationrequest_controller.go

```diff
@@ -47,16 +67,242 @@ type OperationRequestReconciler struct {
+		if err := r.deleteRequest(ctx, request); err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+	// return early if its status is isCompleted and didn't exist for retention time
+	if r.isCompleted(request) {
+		r.Logger.V(1).Info("request is completed and requeue", getLog(request)...)
+		return ctrl.Result{RequeueAfter: OperationReqRequeueDuration}, nil
+	}
+	// change OperationRequest status to failed if it is expired
+	if r.isExpired(request) {
+		r.Logger.V(1).Info("request is expired, update status to failed", getLog(request)...)
+		if err := r.updateRequestStatus(ctx, request, devboxv1alpha1.RequestFailed); err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+	// update OperationRequest status to processing
+	err := r.updateRequestStatus(ctx, request, devboxv1alpha1.RequestProcessing)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	devbox := &devboxv1alpha1.Devbox{}
+	devboxInfo := types.NamespacedName{
+		Name:      request.Spec.DevBoxName,
+		Namespace: request.Namespace,
+	}
+	if err := r.Get(ctx, devboxInfo, devbox); err != nil {
+		return ctrl.Result{}, err
+	}
+	devboxPod := &corev1.Pod{}
+	err = r.Get(ctx, client.ObjectKey{Namespace: devbox.Namespace, Name: devbox.Name}, devboxPod)
+	switch request.Spec.Action {
+	case devboxv1alpha1.Open:
+		if err != nil && client.IgnoreNotFound(err) != nil {
+			// error other than not found
+			logger.Error(err, "get devbox pod failed")
+			return ctrl.Result{}, err
+		} else if err != nil && client.IgnoreNotFound(err) == nil {
+			// no devbox pod found, create a new one
+			err = r.CreatePod(ctx, devbox, req)
+		} else if err == nil {
+			// no need to create pod if it already exists and is running or pending
+			if devboxPod.Status.Phase == corev1.PodRunning || devboxPod.Status.Phase == corev1.PodPending {
+				return ctrl.Result{}, err
+			}
+		}
+	case devboxv1alpha1.Close:
+		if err != nil && client.IgnoreNotFound(err) != nil {
+			// error other than not found
+			logger.Error(err, "get devbox pod failed")
+			return ctrl.Result{}, err
+		} else if err != nil && client.IgnoreNotFound(err) == nil {
+			// no devbox pod found, do nothing
+		} else if err == nil {
+			// no need to create pod if it already exists and is running or pending
+			if devboxPod.Status.Phase != corev1.PodSucceeded {
+				err = r.Delete(ctx, devboxPod)
+				if err != nil {
+					logger.Error(err, "delete devbox pod failed")
+					return ctrl.Result{}, err
+				}
+			}
+		}
+	}
 	// TODO(user): your logic here
 
 	return ctrl.Result{}, nil
 }
+func (r *OperationRequestReconciler) generateImageName(devbox *devboxv1alpha1.Devbox) string {
+	now := time.Now()
+	return fmt.Sprintf("%s/%s/%s:%s", r.CommitImageRegistry, devbox.Namespace, devbox.Name, now.Format("2006-01-02-150405"))
+}
+
+func (r *OperationRequestReconciler) CreatePod(ctx context.Context, devbox *devboxv1alpha1.Devbox, req ctrl.Request) error {
```

Comment:
why do you need create pod? only update devbox spec to running and devbox controller will create pod.